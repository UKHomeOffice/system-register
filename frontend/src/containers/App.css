:root {
  --red: #d62728;
  --risk-colour-high: #d62728;
  --risk-colour-mediumHigh: #ff7f0e;
  --risk-colour-medium: #F9C12C;
  --risk-colour-mediumLow: #BCBD21;
  --risk-colour-low: #2ca02c;
  --risk-colour-unknown: #7f7f7f;
  --blue: #007385;
}

div.topMargin {
  margin-top: 20px;
}

div.topMarginBig {
  margin-top: 40px;
}

div.bottomMargin {
  margin-bottom: 20px;
}

.centerContent {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}

.highlight {
  background-color: var(--blue);
  color: white;
  border-radius: 5px;
  padding: 4px;
}

.badge {
  color: white;
  font-weight: bold;
  padding: 3px;
  padding-right: 4px;
  padding-left: 4px;
  border-radius: 4px;
  margin-right: 4px;
}

.blue {
  background-color: var(--blue);
}

rect {
  transition: all 0.2s ease-in;
}

.notApplicableRisk {
  color: var(--risk-colour-blue);
}
rect.notApplicableRisk {
  fill: var(--risk-colour-blue);
}
.notApplicableRiskBackground {
  background-color: var(--blue);
  color: white;
}

.unknownRisk {
  color: var(--risk-colour-unknown);
}
rect.unknownRisk {
  fill: var(--risk-colour-unknown);
}
.unknownRiskBackground {
  background-color: var(--risk-colour-unknown);
  color: white;
}

.lowRisk{
  color: var(--risk-colour-low);
}
rect.lowRisk{
  fill: var(--risk-colour-low);
}
.lowRiskBackground {
  background-color: var(--risk-colour-low);
  color: white;
}

.mediumLowRisk{
  color: var(--risk-colour-mediumLow);
}
rect.mediumLowRisk{
  fill: var(--risk-colour-mediumLow);
}
.mediumLowRiskBackground {
  background-color: var(--risk-colour-mediumLow);
  color: white;
}

.mediumRisk{
  color: var(--risk-colour-medium);
}
rect.mediumRisk{
  fill: var(--risk-colour-medium);
}
.mediumRiskBackground {
  background-color: var(--risk-colour-medium);
  color: white;
}

.mediumHighRisk{
  color: var(--risk-colour-mediumHigh);
}
rect.mediumHighRisk{
  fill: var(--risk-colour-mediumHigh);
}
.mediumHighRiskBackground {
  background-color: var(--risk-colour-mediumHigh);
  color: white;
}

.highRisk {
  color: var(--risk-colour-high);
}
rect.highRisk{
  fill: var(--risk-colour-high);
}
.highRiskBackground {
  background-color: var(--risk-colour-high);
  color: white;
}

rect.blue1 {
  fill: #eff3ff;
}

rect.blue2 {
  fill: #bdd7e7;
}

rect.blue3 {
  fill: #6baed6;
}

rect.blue4 {
  fill: #3182bd;
}

rect.blue5 {
  fill: #08519c;
}
