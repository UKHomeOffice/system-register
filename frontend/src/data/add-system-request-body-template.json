{
  "portfolio": "Unknown",
  "criticality": null,
  "investment_state": null,
  "business_owner": null,
  "service_owner": null,
  "tech_owner": null,
  "product_owner": null,
  "information_asset_owner": null,
  "developed_by": null,
  "supported_by": null,
  "risks": [
    {
      "name": "architecture",
      "level": null,
      "rationale": "This risk has not been assessed."
    },
    {
      "name": "change",
      "level": null,
      "rationale": "This risk has not been assessed."
    },
    {
      "name": "commercial",
      "level": null,
      "rationale": "This risk has not been assessed."
    },
    {
      "name": "technology",
      "level": null,
      "rationale": "This risk has not been assessed."
    }
  ]
}
