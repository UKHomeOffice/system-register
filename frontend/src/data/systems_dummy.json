{
    "systems": [
        {
            "aliases": [
                "systems register",
                "systems audit",
                "system audit"
            ],
            "businessOwner": null,
            "criticality": "low",
            "description": "Central source of system names, contacts and risk information",
            "developedBy": "Home Office",
            "informationAssetOwner": null,
            "investmentState": "evergreen",
            "name": "System Register",
            "portfolio": "SPAN-R",
            "productOwner": null,
            "risks": [
                {
                    "level": "low",
                    "name": "change",
                    "rationale": "Designed to be easy to change"
                }
            ],
            "serviceOwner": "Home Office",
            "supportedBy": "All",
            "technicalOwner": "Home Office",
            "id": 1,
            "last_updated": {
                "timestamp": "2020-11-18T11:22:50.326570Z",
                "author_name": "Betty Franklin"
            }
        },
        {
            "aliases": [
                "testing more entries"
            ],
            "businessOwner": null,
            "criticality": "high",
            "description": "I'm a cute new system",
            "developedBy": "Home Office",
            "informationAssetOwner": null,
            "investmentState": "sunset",
            "name": "Our second system",
            "portfolio": "SPAN-R",
            "productOwner": null,
            "risks": [
                {
                    "level": "medium",
                    "name": "scorpion",
                    "rationale": "Designed to sting"
                }
            ],
            "serviceOwner": "Home Office",
            "supportedBy": "All",
            "technicalOwner": "Fun Office",
            "id": 2,
            "last_updated": {
                "timestamp": "2020-11-18T11:22:50.326570Z",
                "author_name": "Betty Franklin"
            }
        }
    ],
    "timestamp": "2020-10-09T16:04:36.227867Z"
}