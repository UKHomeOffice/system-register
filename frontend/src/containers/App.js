import React from 'react'
import './App.css';
import SystemList from '../components/SystemList/SystemList'
import TitleBar from '../components/TitleBar/TitleBar'
import { BrowserRouter, Route, Switch } from 'react-router-dom'
import System from '../components/System/System'
import Menu from '../components/Menu/Menu'
import PortfolioHeatmap from '../components/Visualisations/PortfolioHeatmap/PortfolioHeatmap';
import { KeycloakProvider } from '@react-keycloak/web'
import keycloak from '../utilities/keycloak'
import axios from 'axios'
import config from '../config/config'
import Banner from '../components/Banner/Banner'
import AnchorLink from '@govuk-react/link'

// todo state should be in containers, components should be stateless
class App extends React.Component {
    state = {
        register: {
            "systems": []
        }
    }

    componentDidMount() {
        this._isMounted = true;
        axios.get(`${config.api.url}/systems`)
            .then(res => {
                if (this._isMounted) {
                    this.setState({ register: res.data });
                }
            })
            .catch(e => {
                console.error("Error calling api, displaying dummy data!: " + e)
            })
    }

    componentWillUnmount() {
        this._isMounted = false;
    }

    render() {
        return (
            <KeycloakProvider keycloak={keycloak}>
                <BrowserRouter>
                    <header>
                        <TitleBar />
                        <Banner phase="in development">
                            This is a new service - your <AnchorLink href="/contact">feedback</AnchorLink> will help us to improve it.
                    </Banner>
                    </header>
                    <Menu />
                    <main>
                        <Switch>
                            <Route exact path="/">
                                <SystemList register={this.state.register} />
                            </Route>
                            <Route exact path="/system/:id" render={({ match }) => {
                                return <System
                                    system={this.state.register?.systems.find(s => s.id.toString() === match.params.id)} />
                            }}>
                            </Route>
                            <Route exact path="/risk_dashboard">
                                <PortfolioHeatmap systems={this.state.register?.systems} />
                            </Route>
                        </Switch>
                    </main>
                </BrowserRouter>
            </KeycloakProvider>
        );
    }
}

export default App;
