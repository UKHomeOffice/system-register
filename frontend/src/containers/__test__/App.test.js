import React from 'react';
import App from '../App';
import { Router } from 'react-router-dom'
import { render, waitFor, waitForElementToBeRemoved } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { createMemoryHistory } from 'history'
import { rest } from "msw";
import { setupServer } from "msw/node";

import data from '../../data/systems_dummy.json';

const server = setupServer(
  rest.get("/api/systems", (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json(data)
    );
  }),
  rest.get("/config/keycloak", (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json({
        host: "http://localhost:8081/auth/",
        realm: "local-realm",
        clientId: "system-register",
      })
    );
  })
);

describe("<App />", () => {
  beforeAll(() => server.listen({ onUnhandledRequest: "error" }));

  afterAll(() => server.close());

  afterEach(() => server.resetHandlers());

  it('renders system from the API', async () => {
    const history = createMemoryHistory()
    const { findByText, getByTestId } = render(
      <Router history={history}>
        <App />
      </Router>
    )
    await waitForElementToBeRemoved(getByTestId(/auth-initialising-msg/))

    expect(await findByText('System Register', { selector: '.systemCard *' })).toBeInTheDocument();
  })

  it("re-renders from api when a system calls back to notify its been updated", () => {
    //TODO: write this test (JB)
  })

  describe("routing", () => {
    it('navigates to the risk dashboard', async () => {
      const history = createMemoryHistory();
      const { getByText, getByTestId } = render(
        <Router history={history}>
          <App />
        </Router>
      )
      await waitForElementToBeRemoved(getByTestId(/auth-initialising-msg/))
      userEvent.click(getByText(/Risk Dashboard/i))

      await waitFor(() => {
        const dashboard = getByText('Aggregated risk by portfolio')
        expect(dashboard).toBeInTheDocument()
      })
    });

    it("navigates to the About page", async () => {
      const history = createMemoryHistory();
      const { findByRole, findByText, getByTestId } = render(
        <Router history={history}>
          <App />
        </Router>
      );
      await waitForElementToBeRemoved(getByTestId("auth-initialising-msg"))
      const aboutLink = await findByText("About", { selector: "nav a" });

      userEvent.click(aboutLink);

      expect(await findByRole("heading", { level: 1 })).toHaveTextContent("About the System Register");
    });

    it("navigates to the Contact page", async () => {
      const history = createMemoryHistory();
      const { findByRole, findByText, getByTestId } = render(
        <Router history={history}>
          <App />
        </Router>
      );
      await waitForElementToBeRemoved(getByTestId("auth-initialising-msg"))
      const contactLink = await findByText("Contact", { selector: "nav a" });

      userEvent.click(contactLink);

      expect(await findByRole("heading", { level: 1 })).toHaveTextContent("Get in touch");
    });
  })
})
