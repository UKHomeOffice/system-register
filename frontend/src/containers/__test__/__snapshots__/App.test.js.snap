// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `
Array [
  <div
    className="title bottomMargin"
    data-testid="title-bar-div"
  >
    <h2
      className="title"
    >
      System Register
    </h2>
    <div
      className="auth"
    >
      <button
        className="login"
        href=""
        onClick={[Function]}
      >
        Login
      </button>
    </div>
  </div>,
  <div
    className="Menu"
  >
    <ul>
      <li>
        <a
          aria-current={null}
          href="/"
          onClick={[Function]}
        >
          System Register
        </a>
      </li>
      <li>
        <a
          aria-current="page"
          className="selected"
          href="/risk_dashboard"
          onClick={[Function]}
          style={Object {}}
        >
          Risk Dashboard
        </a>
      </li>
      <li>
        <a
          aria-current={null}
          href="/about"
          onClick={[Function]}
        >
          About
        </a>
      </li>
      <li>
        <a
          aria-current={null}
          href="/contact"
          onClick={[Function]}
        >
          Contact
        </a>
      </li>
    </ul>
  </div>,
  <div
    className="portfolioHeatmapContainer"
  >
    <div
      className="centerContent"
    >
      <h2>
        Aggregated risk by portfolio
      </h2>
      <p>
        The following visulisation shows an aggregated view of risk across each portfolio. Risk is broken down by different "lenses" such as roadmap, technology or commercial. Each system is individualy scored against each of these lenses, and the scores of all systems within a portfolio are summed up to provide the overall scores visible below.
      </p>
      <p>
        These scores are normalised against the number of systems within the portfolio, so portfolios with many systems can be compared more easily against portfolios with fewer.
      </p>
    </div>
    <div
      className="visContainer"
    >
      <div
        className="leftVis"
      >
        <div
          className="systemsHeatmap"
        />
        <fieldset
          className="MuiFormControl-root"
        >
          <div
            className="MuiFormGroup-root MuiFormGroup-row"
            role="radiogroup"
          >
            <label
              aria-describedby={null}
              className="MuiFormControlLabel-root"
              data-testid="radioKnownRisk"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseLeave={[Function]}
              onMouseOver={[Function]}
              onTouchEnd={[Function]}
              onTouchStart={[Function]}
              title="Display the scores of all the systems that have been audited.
    Red = higher risk, green = lower risk."
            >
              <span
                aria-disabled={false}
                className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorSecondary PrivateSwitchBase-checked-2 Mui-checked MuiIconButton-colorSecondary"
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={null}
              >
                <span
                  className="MuiIconButton-label"
                >
                  <input
                    checked={true}
                    className="PrivateSwitchBase-input-4"
                    disabled={false}
                    name="view"
                    onChange={[Function]}
                    type="radio"
                    value="knownRisk"
                  />
                  <div
                    className="PrivateRadioButtonIcon-root-5 PrivateRadioButtonIcon-checked-7"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                      />
                    </svg>
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                      />
                    </svg>
                  </div>
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </span>
              <span
                className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
              >
                Known risk only
              </span>
            </label>
            <label
              aria-describedby={null}
              className="MuiFormControlLabel-root"
              data-testid="radioUnknownRisk"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseLeave={[Function]}
              onMouseOver={[Function]}
              onTouchEnd={[Function]}
              onTouchStart={[Function]}
              title="Display scores that treat unknown risk
    (including systems that have not been audited) as inherantly risky.
    Red = higher risk, green = lower risk."
            >
              <span
                aria-disabled={false}
                className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={null}
              >
                <span
                  className="MuiIconButton-label"
                >
                  <input
                    checked={false}
                    className="PrivateSwitchBase-input-4"
                    disabled={false}
                    name="view"
                    onChange={[Function]}
                    type="radio"
                    value="includeUnknownRisk"
                  />
                  <div
                    className="PrivateRadioButtonIcon-root-5"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                      />
                    </svg>
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                      />
                    </svg>
                  </div>
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </span>
              <span
                className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
              >
                Include unknown risk
              </span>
            </label>
            <label
              aria-describedby={null}
              className="MuiFormControlLabel-root"
              data-testid="radioCoverage"
              onBlur={[Function]}
              onFocus={[Function]}
              onMouseLeave={[Function]}
              onMouseOver={[Function]}
              onTouchEnd={[Function]}
              onTouchStart={[Function]}
              title="Display the proportion of known systems have been audited against each portfolio risk lense.
    Blue = more coverage, White = less coverage."
            >
              <span
                aria-disabled={false}
                className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiRadio-root MuiRadio-colorSecondary MuiIconButton-colorSecondary"
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={null}
              >
                <span
                  className="MuiIconButton-label"
                >
                  <input
                    checked={false}
                    className="PrivateSwitchBase-input-4"
                    disabled={false}
                    name="view"
                    onChange={[Function]}
                    type="radio"
                    value="unknowns"
                  />
                  <div
                    className="PrivateRadioButtonIcon-root-5"
                  >
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                      />
                    </svg>
                    <svg
                      aria-hidden={true}
                      className="MuiSvgIcon-root PrivateRadioButtonIcon-layer-6"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                      />
                    </svg>
                  </div>
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </span>
              <span
                className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
              >
                Audit coverage
              </span>
            </label>
          </div>
        </fieldset>
        <div
          className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
        >
          <svg
            height={200}
            id="summaryVis"
            width={580}
          />
        </div>
      </div>
      <div
        className="rightVis"
      >
        <h3>
          Combined
           risk to systems in 
          Home Office
           
          
        </h3>
        <div
          className="riskList"
        >
          <div
            className="MuiPaper-root MuiTableContainer-root makeStyles-container-10 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <table
              aria-label="a dense table"
              className="MuiTable-root makeStyles-table-9 MuiTable-stickyHeader"
              role={null}
            >
              <thead
                className="MuiTableHead-root"
                role={null}
              >
                <tr
                  className="MuiTableRow-root MuiTableRow-head"
                  role={null}
                >
                  <th
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-head makeStyles-head-8 MuiTableCell-alignLeft MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
                    scope="col"
                  >
                    <strong>
                      System
                    </strong>
                  </th>
                  <th
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-head makeStyles-head-8 MuiTableCell-alignLeft MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
                    scope="col"
                  >
                    <strong>
                      Criticality
                    </strong>
                  </th>
                  <th
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-head makeStyles-head-8 MuiTableCell-alignCenter MuiTableCell-sizeSmall MuiTableCell-stickyHeader"
                    scope="col"
                  >
                    <strong>
                      Combined
                       Risk
                    </strong>
                  </th>
                </tr>
              </thead>
              <tbody
                className="MuiTableBody-root"
                role={null}
              >
                <tr
                  className="MuiTableRow-root"
                  role={null}
                >
                  <td
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeSmall"
                    scope="row"
                  >
                    <a
                      href="/system/1"
                      onClick={[Function]}
                    >
                      System Register
                    </a>
                  </td>
                  <td
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignLeft MuiTableCell-sizeSmall"
                  >
                    <span
                      className="badge criticalityLow"
                    >
                      
                      Low
                    </span>
                  </td>
                  <td
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-body lowRiskBackground MuiTableCell-alignLeft MuiTableCell-sizeSmall"
                    data-testid="risk_cell_System Register"
                  >
                    <strong
                      style={
                        Object {
                          "color": "white",
                        }
                      }
                    >
                      0
                    </strong>
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
